### DISCUSSION TITLE
The right way to implement SQL-based working memory?

### DISCUSSION THREAD

**Discussion by Jeshuakrc** [evrete/evrete#61] [Q&A]
Hey!
For my project, I need rulesets to be able to access the database as its working memory. Eventually I'll implement a custom DSL, bur as of now, I'm relaying on the default SPI implementations and the Java-DSL to author rules. 

So far, I use Spring's Event API to create and fire sessions, but I'm manually inserting facts into the session before firing them. Rules should have access to the whole database, where facts are queried as JPA entities.

It's clear to me that I need to extend Evrete using Hibernate to make all entities accessible as facts to the RuntimeContext, but I've been exploring the API and the Core code, and the path isn't clear to me. I am not sure if I should implement a custom DSLKnowledgeProvider, a custom TypeResolverProvider or a MemoryFactoryProvider.

I'd appreciate any advice!

**Comment by andbi** [comment]
Hi!

There's no way around manually inserting initial facts and firing sessions, but nothing prevents you from dealing with facts that are managed by an `EntityManager`. 

Moreover, the `EntityManager` itself can be stored as an environment variable and accessed in your rule actions. So, for example, querying your database and inserting new facts from a rule's RHS shouldn't be a problem at all.

Creating a new `DSLKnowledgeProvider` is an overkill imho, but you might want to implement the `MemoryFactoryProvider` later if you decide to keep all your facts in the database instead of the engine's working memory (RAM).

Regards,
Andrew
